### ISSUE TITLE
[FEATURE] Rewiring `cumsum` via (parts of) the grid_ufunc logic.

### ISSUE THREAD

**Issue by jbusecke** [xgcm/xgcm#458]
While working on https://github.com/jbusecke/xgcm/tree/vector_padding I am encountering some test failures, specifically the `xgcm/test/test_grid.py::test_keep_coords[xx-cumint]`, `xgcm/test/test_grid.py::test_keep_coords[xx-cumsum]` tests which are testing the consistency of output coordinates. 

While checking on this I noticed that the `Grid.cumsum` method in the `grid_ufunc_refactor` branch still calls the `Axis.cumsum` method, which has a bunch of [custom logic for padding/rewrapping the output](https://github.com/jbusecke/xgcm/blob/c742c47dbfee0a48b034baec965d33ef3fb21579/xgcm/grid.py#L702-L721) ([which itself is simply xarray.cumsum()](https://github.com/jbusecke/xgcm/blob/c742c47dbfee0a48b034baec965d33ef3fb21579/xgcm/grid.py#L700)). 
We will ultimately have to refactor this logic into the Grid method, when we remove the Axis methods. 

Should/Can this logic be implemented via the gufunc logic? That would solve my issue here, since [these changes I made](https://github.com/jbusecke/xgcm/blob/c742c47dbfee0a48b034baec965d33ef3fb21579/xgcm/grid_ufunc.py#L994-L1026) would ensure that the treatment of output coordinates is always consistent.

**Comment by TomNicholas** [comment]
My plan is to add an extra `pad_before` kwarg to `apply_as_grid_ufunc`, which defaults to `True`. If it's set to false then we instead pad *after* the `xr.apply_ufunc` call. Then we have various `cumsum` ufuncs defined in `gridops.py` that have different `boundary_width` and trimming behaviour to replicate the bespoke `_pad_array` logic that @rabernat originally wrote.

This approach has the advantage of being easy to change if we do end up [moving padding inside `xr.apply_ufunc`](https://github.com/xgcm/xgcm/issues/436) later: it would become just a flip in the order of the function composition. It also means that someone could define new grid ufuncs which pad after their application (I cannot think of any examples of this other than cumulative sums/integrals, but we would get this option for free).

**Comment by TomNicholas** [comment]
The other complication is that we don't want to use `map_overlap=True` for chunked core dims for the specific case of cumsum, because unlike all the other pre-defined ufuncs it handles dask arrays itself

**Comment by TomNicholas** [comment]
This set of grid ufuncs to go in `gridops.py`

```python
# Cumsum

@as_grid_ufunc(signature="(X:center)->(X:left)", boundary_width={"X": (1, 0)}, fill_value=0)
def cumsum_center_to_left(a):
    # TODO will this dispatch to dask.array.cumsum if it encounters a dask array?
    return np.cumsum(a, axis=-1)[..., :-1]


@as_grid_ufunc(signature="(X:left)->(X:center)", boundary_width={"X": (0, 0)})
def cumsum_left_to_center(a):
    return np.cumsum(a, axis=-1)


@as_grid_ufunc(signature="(X:center)->(X:right)", boundary_width={"X": (0, 0)})
def cumsum_center_to_right(a):
    # TODO will this dispatch to dask.array.cumsum if it encounters a dask array?
    return np.cumsum(a, axis=-1)


@as_grid_ufunc(signature="(X:right)->(X:center)", boundary_width={"X": (1, 0)}, fill_value=0)
def cumsum_right_to_center(a):
    # TODO will this dispatch to dask.array.cumsum if it encounters a dask array?
    return np.cumsum(a, axis=-1)[..., :-1]


@as_grid_ufunc(signature="(X:center)->(X:outer)", boundary_width={"X": (1, 0)}, fill_value=0)
def cumsum_center_to_outer(a):
    return np.cumsum(a, axis=-1)


@as_grid_ufunc(signature="(X:outer)->(X:center)", boundary_width={"X": (0, 0)})
def cumsum_outer_to_center(a):
    return np.cumsum(a, axis=-1)[..., :-1]


@as_grid_ufunc(signature="(X:center)->(X:inner)", boundary_width={"X": (0, 0)})
def cumsum_center_to_inner(a):
    return np.cumsum(a, axis=-1)[..., :-1]


@as_grid_ufunc(signature="(X:inner)->(X:center)", boundary_width={"X": (1, 0)}, fill_value=0)
def cumsum_inner_to_center(a):
    return np.cumsum(a, axis=-1)
```

should hopefully represent this older code

```python
# now pad / trim the data as necessary
# here we enumerate all the valid possible shifts
if (pos == "center" and to == "right") or (pos == "left" and to == "center"):
    # do nothing, this is the default for how cumsum works
    data = da_cum.data
elif (pos == "center" and to == "left") or (pos == "right" and to == "center"):
    data = _pad_array(
        da_cum.isel(**{dim: slice(0, -1)}), dim, left=True, **boundary_kwargs
    )
elif (pos == "center" and to == "inner") or (pos == "outer" and to == "center"):
    data = da_cum.isel(**{dim: slice(0, -1)}).data
elif (pos == "center" and to == "outer") or (pos == "inner" and to == "center"):
    data = _pad_array(da_cum, dim, left=True, **boundary_kwargs)
else:
    raise ValueError(
```

**Comment by TomNicholas** [comment]
This was closed by #461
